bundle/src/main/java/com/adobe/acs/commons/mcp/impl/DialogResourceProviderFactoryImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * #%L
+ * ACS AEM Commons Bundle
+ * %%
+ * Copyright (C) 2019 Adobe
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.mcp.impl;
+
+import com.adobe.acs.commons.mcp.DialogResourceProviderConfiguration;
+import com.adobe.acs.commons.mcp.DialogResourceProviderFactory;
+import com.adobe.acs.commons.mcp.form.DialogProvider;
+import java.util.Collections;
+import java.util.Dictionary;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+import org.apache.commons.lang3.ClassUtils;
+import org.apache.sling.api.adapter.AdapterFactory;
+import org.apache.sling.spi.resource.provider.ResourceProvider;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+import org.osgi.service.metatype.annotations.Designate;
+import org.slf4j.LoggerFactory;
+
+@Component(
+        service = DialogResourceProviderFactory.class,
+        immediate = true
```

Comment:
wouldn't it be better to make this as configurationPolicy=required rather than using an explicit enabled flag?  It sorta does the same thing, but that would disable it at an osgi framework level instead of a code level.

Reply:
Well, no in this case I need the service aware of things that are registered so I need the service "running" but when it's not enabled it just doesn't create or register the resource providers.  This kind of operation also allows it to work properly when the OSGi service is reconfigured, such as if it is enabled later in the lifecycle via configuration after several sling models have already been registered as AdaperFactory instances.   Since catching registration of sling models is already really a weird thing, I'm more comfortable leaving this one as it is.  For what it's worth, there is a unit test that ensures disable/enable of the service works properly.

Reply:
Good call though @shsteimer  -- in a normal OSGi service I'd be happy with that, but this one is a slight step into "weird" territory and there's only so much that OSGi mocks let me test, so this way is probably safer.

Reply:
That makes sense.  I am now of course curious and want to find time to play with it to see if I can make it work without that need, but if you say it's required that's good enough for me.